File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.cpp
class Solution {
  public:
    int countAtMostK(vector<int> &arr, int k) {
        // code here
        int n=arr.size();
        int i=0;
        int j=0;
        int sum=0;
        unordered_map<int,int>mp;
        while(i<n)
        {
          
          mp[arr[i]]++;
          while(j<i && mp.size()>k)
          {
              --mp[arr[j]];
              if(mp[arr[j]]==0)
              {
                  mp.erase(arr[j]);
              }
              j++;
              
          }
          sum+=j;
          i++;
          
        }
        long long ans=1ll*n*(n+1)/2;
        return ans-sum;
    }
};

File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.java
class Solution {
    public int countAtMostK(int arr[], int k) {
        // code here
        Map<Integer,Integer>mp=new HashMap<>();
        int n=arr.length;
        int j=0;
        int ans=0;
        for(int i=0;i<n;i++)
        {  if(mp.get(arr[i])==null)
        {
            mp.put(arr[i],1);
        }
        else{
            mp.put(arr[i],mp.get(arr[i])+1);
            
        }
            while(j<=i&& mp.size()>k)
            {
                mp.put(arr[j],mp.get(arr[j])-1);
                if(mp.get(arr[j]).equals(0))
                {
                    mp.remove(arr[j]);
                }
                j++;
                
            }
            ans+=(i-j+1);
            
            
        }
        return ans;
        
        
    }
}
